core/chain_indexer.go

```diff
@@ -201,16 +203,27 @@ func (c *ChainIndexer) eventLoop(currentHeader *types.Header, events chan ChainH
 				return
 			}
 			header := ev.Block.Header()
+			addressOutpoints := rawdb.ReadAddressOutpoints(c.chainDb, config.Location)
+
 			if header.ParentHash(nodeCtx) != prevHash {
 				// Reorg to the common ancestor if needed (might not exist in light sync mode, skip reorg then)
 				// TODO: This seems a bit brittle, can we detect this case explicitly?
 
 				if rawdb.ReadCanonicalHash(c.chainDb, prevHeader.NumberU64(nodeCtx)) != prevHash {
-					if h := rawdb.FindCommonAncestor(c.chainDb, prevHeader, header, nodeCtx); h != nil {
+					if h := rawdb.FindCommonAncestor(c.chainDb, prevHeader, header, addressOutpoints, nodeCtx); h != nil {
+
+						collectHeaders := make([]*types.Header, 0)
+						for curr := header; curr != h; curr = rawdb.ReadHeader(c.chainDb, curr.ParentHash(nodeCtx), uint64(nodeCtx)) {
+							collectHeaders = append(collectHeaders, h)
+						}
+
+						c.reorgUtxoIndexer(collectHeaders, addressOutpoints, nodeCtx, config)
 						c.newHead(h.NumberU64(nodeCtx), true)
 					}
 				}
 			}
+
+			c.addOutpointsToIndexer(addressOutpoints, config, ev.Block)
```

Comment:
You should also delete utxos that are no longer valid

Reply:
That's in reorgUtxoIndexer

Reply:
I mean if a block is being removed from the chain (reorg) you should remove all those utxos from the indexer

Reply:
You must go from the common ancestor to the new block header and see if there are blocks in between. If there are, you must add those UTXOs to the indexer